### PR TITLE
ES2015ify and make us of promises

### LINKED ISSUE #203: ES2015 and Promises
I think we should go all in on this module and use ES2015 and Promises. Need to find a way to easily compile with Babel without too much boilerplate. Promise polyfill: https://github.com/floatdrop/pinkie-promise

PR welcome if anyone wants to try their hand at this.

### PR DESCRIPTION
**Pull request by kevva**
Tests fails (at least locally) on the renaming because of the seconds in the name (differs 1-2 secs). Also, used `export fn(args)` instead of `export const fn = (args) =>` because the latter lost context for some reason.

Fixes #203.

### CODE REVIEW

File: test/test.js [hunk 1]
```diff
@@ -41,226 +35,215 @@ test('error if no url is specified', t => {
...
 
-	pageres.run((err, streams) => {
-		t.ifError(err);
-		t.is(streams.length, 1);
-		t.is(streams[0].filename, 'microsoft.com!query=pageres-1024x768.png');
-	});
+	new Pageres()
+		.src('http://www.microsoft.com/?query=pageres*|<>:"\\', ['1024x768'])
+		.run()
+		.then(streams => {
```
Comment (sindresorhus) [inline test/test.js]: Happy to have this done in a follow-up PR if you prefer, but would be super nice to use async/await in the tests.

https://github.com/sindresorhus/temp-write/blob/985ea544148bdcd15df4d5e960ddb6c0772c4194/test.js#L7

File: lib/index.js [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+					this.resolution(src.url, options).then(resolve);
+					return;
+				}
+
+				if (keywords.length) {
+					this.viewport({url: src.url, sizes, keywords}, options).then(resolve);
+					return;
+				}
+
+				sizes.forEach(size => {
```
Comment (sindresorhus) [inline lib/index.js]: Can use `for/of` here.

File: lib/index.js [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+					this.sizes.push(size);
+
+					try {
```
Comment (sindresorhus) [inline lib/index.js]: Don't need to try/catch now that it's in a promise.

File: lib/util.js [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+				reject(err);
+				return;
+			}
+
+			items.forEach(item => {
+				this.sizes.push(item.size);
+				obj.sizes.push(item.size);
+			});
+
+			arrayUniq(obj.sizes).forEach(size => {
```
Comment (sindresorhus) [inline lib/util.js]: `for/of` and no need for try/catch.

File: lib/util.js [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+ *
+ * @param {Object} obj
+ * @param {Object} options
+ */
+
+export function viewport(obj, options) {
+	const v = memoizeAsync(viewportList);
+
+	return new Promise((resolve, reject) => {
+		v(obj.keywords, (err, items) => {
```
Comment (sindresorhus) [inline lib/util.js]: Use `pify`

File: readme.md [hunk 6]
```diff
@@ -168,7 +166,7 @@ Set the destination directory.
 
 Type: `string`
 
-### pageres.run(callback)
+### pageres.run()
 
 Run pageres.
```
Comment (sindresorhus) [inline readme.md]: You need to remove the below `callback` section.

File: readme.md [hunk 7]
```diff
@@ -168,7 +166,7 @@ Set the destination directory.
...
```
Comment (sindresorhus) [inline readme.md]: Mention that it returns a promise and what that promise resolves to.

**Comment by sindresorhus** [comment]
// @arthurvr @Qix- @vdemedes @SamVerschueren in case any of you would like to help review :)

**Comment by sindresorhus** [comment]
> Also, used export fn(args) instead of export const fn = (args) => because the latter lost context for some reason.

Yes, that's fine, arrow function doesn't have `this` context.

File: lib/util.js [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+			}
+
+			items.forEach(item => {
+				this.sizes.push(item.item);
+
+				try {
+					this.items.push(this.create(url, item.item, options));
+				} catch (err) {
+					reject(err);
+				}
```
Comment (SamVerschueren) [inline lib/util.js]: No need for try-catch I guess as you are in a promise already.

File: test/test.js (hunk 1 above)
Comment (SamVerschueren) [inline test/test.js]: Would make the tests much cleaner indeed. I you don't find the time, would be happy to do it as well.

File: lib/index.js (hunk 3 above)
Comment (kevva) [inline lib/index.js]: Oh, yeah, true!

File: test/serverForCookieTests.js [hunk 9]
```diff
@@ -1,15 +1,15 @@
 'use strict';
```
Comment (SamVerschueren) [inline test/serverForCookieTests.js]: Is `use strict` still necessary?

File: test/test.js (hunk 1 above)
Comment (kevva) [inline test/test.js]: Yup, but I rather do it in a separate PR.

File: test/serverForCookieTests.js (hunk 9 above)
Comment (kevva) [inline test/serverForCookieTests.js]: Nope, good catch!

**Comment by kevva** [comment]
> No need to pass in Promise as pify already uses pinkie-promise.

Oh, it does? Haven't had a chance to look it over yet, but good to know.

**Comment by SamVerschueren** [comment]
Oeps, sorry. I checked it again and it seems it doesn't. My bad!

**Comment by sindresorhus** [comment]
No, you have to pass in `pinkie-promise` manually. I didn't want `pify` to be tied to any specific promise lib.

**Comment by sindresorhus** [comment]
> Tests fails (at least locally) on the renaming because of the seconds in the name (differs 1-2 secs).

Use a regex for the time part.

**Comment by kevva** [comment]
Tests should pass now.

File: lib/index.js [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+					return;
+				}
+
+				for (const size of sizes) {
+					this.sizes.push(size);
+					this.items.push(this.create(src.url, size, options));
+				}
+
+				resolve();
+			});
```
Comment (SamVerschueren) [inline lib/index.js]: Instead of wrapping this entire codeblock in a Promise, maybe it can be written like this?

```js
return Promise.all(this.src().map(src => {
	const options = objectAssign({}, this.options, src.options);
	const sizes = arrayUniq(src.sizes.filter(/./.test, /^\d{3,4}x\d{3,4}$/i));
	const keywords = arrayDiffer(src.sizes, sizes);

	if (!src.url) {
		return Promise.reject(new Error('URL required'));
	}

	this.urls.push(src.url);

	if (!sizes.length && keywords.indexOf('w3counter') !== -1) {
		return this.resolution(src.url, options);
	}

	if (keywords.length) {
		return this.viewport({url: src.url, sizes, keywords}, options);
	}

	for (const size of sizes) {
		this.sizes.push(size);
		this.items.push(this.create(src.url, size, options));
	}

	return Promise.resolve();
}))
```

File: lib/index.js [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+	run() {
+		return Promise.all(this.src().map(src => {
+			return new Promise((resolve, reject) => {
+				const options = objectAssign({}, this.options, src.options);
+				const sizes = arrayUniq(src.sizes.filter(/./.test, /^\d{3,4}x\d{3,4}$/i));
+				const keywords = arrayDiffer(src.sizes, sizes);
+
+				if (!src.url) {
+					reject(new Error('URL required'));
```
Comment (SamVerschueren) [inline lib/index.js]: Return after promise is rejected?

File: lib/index.js (hunk 10 above)
Comment (SamVerschueren) [inline lib/index.js]: As a matter of fact, I don't think the last `Promise.resolve()` is necessary.

Comment (kevva) [inline lib/index.js]: :+1:

**Comment by sindresorhus** [comment]
:heart: The code looks so much cleaner and readable now! :heart: 

Thanks @kevva :)
